Automate Helm tests with Gradle

adutra opened the pull request:
This PR automates Helm tests and make them executable as part of the Gradle build.

singhpk234 commented on helm/polaris/README.md.gotmpl:
````diff
@@ -185,4 +185,13 @@ Integration tests are run with the Chart Testing tool:
 ct install --namespace polaris --charts ./helm/polaris
 ```
 
+### Running tets with Gradle
````
```suggestion
### Running test with Gradle
```

dimas-b reviewed (approved):
LGTM 👍

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+description = "Apache Polaris (incubating) Helm Chart"
+
+val helmTestReportsDir = layout.buildDirectory.dir("reports")
+
```
```suggestion

val runtimeServerDistribution by configurations.registering {
  isCanBeConsumed = true
  isCanBeResolved = false
}

dependencies {
  runtimeServerDistribution(project(":polaris-server", "distributionElements"))
}
```

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+          .trimIndent(),
+      )
+
+    inputs.files(
```
```suggestion
    inputs.files(runtimeServerDistribution)
    inputs.files(
```

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+    workingDir = rootProject.projectDir
+
+    // Output: file containing SHA digest of built image
+    val outputFile = helmTestReportsDir.get().file("minikube-images.sha256")
```
Better relativize `outputFile` to `project.projectDir` (or better `this.workingDir`) here (and in other places) so the `commandLine` doesn't contain any local paths.

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+    dependsOn(":polaris-server:quarkusBuild")
+
+    inputs.dir(rootProject.file("runtime/server/build/quarkus-app"))
+    inputs.file(rootProject.file("runtime/server/src/main/docker/Dockerfile.jvm"))
```
Hm - better use another Gradle configuration pair. Referencing other projects leads to issues.

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+          .trimIndent(),
+      )
+
+    inputs.files(
```
I guess this is needed as well:

```suggestion
    inputs.files(runtimeServerDistribution)
    inputs.files(
```

adutra replied:
This task depends on `buildMinikubeImages` which depends on `:polaris-server:quarkusBuild` – isn't that enough?

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+    group = "verification"
+    description = "Run Helm unit tests"
+
+    workingDir = projectDir
```
```suggestion
```

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+    group = "verification"
+    description = "Run chart-testing (install) on minikube"
+
+    workingDir = rootProject.projectDir
```
Hm, can't it run from the project directory?

snazy commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+      fileTree(projectDir) { include("Chart.yaml", "values.yaml", "ci/**/*", "templates/**/*") }
+    )
+
+    outputs.file(outputFile)
```
All `Exec` tasks also need
```suggestion
    outputs.cacheIf { true }
```
otherwise only up-to-date checks work, but nothing would be cached. (`Exec` is annotated with `@DisableCachingByDefault`.)

snazy replied:
The cached outputs are (hopefully) relatively small-ish, compared to the execution time of these tasks.

snazy reviewed (commented):
This is nice!

Some comments - all Gradle related improvements.

adutra commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,284 @@
+
+    dependsOn(":polaris-server:quarkusBuild")
+
+    inputs.files(runtimeServerDistribution)
```
@snazy I think this is the only task that actually needs this configuration as input.

As for the Dockerfile, I don't see any other way to reference it as it is not an output of any task (that I know of). Should I create a configuration in `polaris-server` to expose it?

adutra commented on helm/polaris/build.gradle.kts:
```diff
@@ -0,0 +1,245 @@
+    group = "build"
+    description = "Build Polaris Docker images in minikube's Docker environment"
+
+    workingDir = rootProject.projectDir
```
This one no, because `docker` cannot read files outside the context root.

adutra commented:
To all the reviewers: sorry for so pushing so many commits 😅 – it was harder than I thought to get this PR right.

Note: with this change, `./gradlew check` or `./gradlew build` would fail if the necessary tools are not installed (helm, ct, minikube...). I think that's fair but happy to discuss this if someone disagrees.

adutra commented:
> Not sure whether we should add all the helm-building/testing dependencies as mandatory build requirements - at least not when running a ./gradlew check or test or so.
> 
> OTOH it would be very convenient for many people to be able to run all those things in a simple way.
> 
> WDYT about having a Dockerfile or docker-compose that includes all the necessary dependencies. Ensuring that writes to the source tree from those containers use the host's user is a bit tricky though.

In this commit I implemented a check for missing dependencies that is entirely shell-based:

https://github.com/apache/polaris/pull/2964/commits/09e638ed63b7b13b2bea0ae40e472a2967d6ce75

But then I removed it, thinking it was a bit overkill. But in hindsight, that was maybe better, wdyt?

adutra commented:
Closing as the idea needs some more polishing.
